The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12660118577/job/35280674206) [(plain)](https://github.com/rust-lang/rust/commit/5f04f98c9aaf04e5f670e04dc12d4665489859fd/checks/35280674206/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 26.9s done
#21 DONE 39.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.79s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
Saved the actual stderr to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/struct-pattern-on-non-struct-resolve-error/struct-pattern-on-non-struct-resolve-error.stderr"
diff of stderr:

20    |            ~~~~~~~~~~~~~~~~~~~~~~~~~~
21 LL |     if let <CharIndices<'_> as Iterator>::Item { .. } = 1 {
-      and 71 other candidates
+      and 65 other candidates
24 
25 error: aborting due to 2 previous errors
---
To only update this specific test, also pass `--test-args pattern/struct-pattern-on-non-struct-resolve-error.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=i686-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/pattern/struct-pattern-on-non-struct-resolve-error" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/i686-unknown-linux-gnu/native/rust-test-helpers" "-Clinker=x86_64-linux-gnu-gcc"
--- stderr -------------------------------
error[E0425]: cannot find value `x` in this scope
##[error]  --> /checkout/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.rs:5:9
   |
   |
LL |         x //~ ERROR E0425

error[E0223]: ambiguous associated type
##[error]  --> /checkout/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.rs:4:12
   |
   |
LL |     if let Iterator::Item { .. } = 1 { //~ ERROR E0223
   |
help: use fully-qualified syntax
   |
   |
LL |     if let <Ancestors<'_> as Iterator>::Item { .. } = 1 { //~ ERROR E0223
   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
LL |     if let <Args as Iterator>::Item { .. } = 1 { //~ ERROR E0223
   |            ~~~~~~~~~~~~~~~~~~~~~~~~
LL |     if let <ArgsOs as Iterator>::Item { .. } = 1 { //~ ERROR E0223
   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~
LL |     if let <CharIndices<'_> as Iterator>::Item { .. } = 1 { //~ ERROR E0223
     and 65 other candidates

error: aborting due to 2 previous errors

```

</details>

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -1130,7 +1130,7 @@ impl<'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'_, 'ast, 'ra, 'tcx> {
         let None = following_seg else { return };
         for rib in self.ribs[ValueNS].iter().rev() {
             for (def_id, spans) in &rib.patterns_with_skipped_bindings {
-                if let DefKind::Struct = self.r.tcx.def_kind(*def_id)
+                if let DefKind::Struct | DefKind::Variant = self.r.tcx.def_kind(*def_id)
                     && let Some(fields) = self.r.field_idents(*def_id)
                 {
                     for field in fields {

--- a/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.rs
+++ b/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.rs
@@ -1,7 +1,10 @@
 // Regression test for #135209.
 // We ensure that we don't try to access fields on a non-struct pattern type.
 fn main() {
-    if let Iterator::Item { .. } = 1 { //~ ERROR E0223
+    if let <Vec<()> as Iterator>::Item { .. } = 1 {
+        //~^ ERROR E0658
+        //~| ERROR E0071
+        //~| ERROR E0277
         x //~ ERROR E0425
     }
 }

--- a/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.stderr
+++ b/tests/ui/pattern/struct-pattern-on-non-struct-resolve-error.stderr
@@ -1,28 +1,34 @@
 error[E0425]: cannot find value `x` in this scope
-  --> $DIR/struct-pattern-on-non-struct-resolve-error.rs:5:9
+  --> $DIR/struct-pattern-on-non-struct-resolve-error.rs:8:9
    |
 LL |         x
    |         ^ not found in this scope
 
-error[E0223]: ambiguous associated type
+error[E0658]: usage of qualified paths in this context is experimental
   --> $DIR/struct-pattern-on-non-struct-resolve-error.rs:4:12
    |
-LL |     if let Iterator::Item { .. } = 1 {
-   |            ^^^^^^^^^^^^^^
+LL |     if let <Vec<()> as Iterator>::Item { .. } = 1 {
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
-help: use fully-qualified syntax
+   = note: see issue #86935 <https://github.com/rust-lang/rust/issues/86935> for more information
+   = help: add `#![feature(more_qualified_paths)]` to the crate attributes to enable
+   = note: this compiler was built on YYYY-MM-DD; consider upgrading it if it is out of date
+
+error[E0071]: expected struct, variant or union type, found inferred type
+  --> $DIR/struct-pattern-on-non-struct-resolve-error.rs:4:12
    |
-LL |     if let <Ancestors<'_> as Iterator>::Item { .. } = 1 {
-   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     if let <Args as Iterator>::Item { .. } = 1 {
-   |            ~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     if let <ArgsOs as Iterator>::Item { .. } = 1 {
-   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~
-LL |     if let <CharIndices<'_> as Iterator>::Item { .. } = 1 {
-   |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-     and 71 other candidates
+LL |     if let <Vec<()> as Iterator>::Item { .. } = 1 {
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a struct
 
-error: aborting due to 2 previous errors
+error[E0277]: `Vec<()>` is not an iterator
+  --> $DIR/struct-pattern-on-non-struct-resolve-error.rs:4:12
+   |
+LL |     if let <Vec<()> as Iterator>::Item { .. } = 1 {
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Vec<()>` is not an iterator
+   |
+   = help: the trait `Iterator` is not implemented for `Vec<()>`
 
-Some errors have detailed explanations: E0223, E0425.
-For more information about an error, try `rustc --explain E0223`.
+error: aborting due to 4 previous errors
+
+Some errors have detailed explanations: E0071, E0277, E0425, E0658.
+For more information about an error, try `rustc --explain E0071`.

--- a/tests/ui/pattern/struct-pattern-with-missing-fields-resolve-error.rs
+++ b/tests/ui/pattern/struct-pattern-with-missing-fields-resolve-error.rs
@@ -3,6 +3,10 @@ struct Website {
     title: Option<String>,
 }
 
+enum Foo {
+    Bar { a: i32 },
+}
+
 fn main() {
     let website = Website {
         url: "http://www.example.com".into(),
@@ -17,5 +21,10 @@ fn main() {
     if let Website { url, .. } = website { //~ NOTE this pattern
         println!("[{}]({})", title, url); //~ ERROR cannot find value `title` in this scope
         //~^ NOTE not found in this scope
+    }
+
+    let x = Foo::Bar { a: 1 };
+    if let Foo::Bar { .. } = x { //~ NOTE this pattern
+        println!("{a}"); //~ ERROR cannot find value `a` in this scope
     }
 }

--- a/tests/ui/pattern/struct-pattern-with-missing-fields-resolve-error.stderr
+++ b/tests/ui/pattern/struct-pattern-with-missing-fields-resolve-error.stderr
@@ -1,5 +1,5 @@
 error: expected `,`
-  --> $DIR/struct-pattern-with-missing-fields-resolve-error.rs:12:31
+  --> $DIR/struct-pattern-with-missing-fields-resolve-error.rs:16:31
    |
 LL |     if let Website { url, Some(title) } = website {
    |            -------            ^
@@ -7,13 +7,21 @@ LL |     if let Website { url, Some(title) } = website {
    |            while parsing the fields for this pattern
 
 error[E0425]: cannot find value `title` in this scope
-  --> $DIR/struct-pattern-with-missing-fields-resolve-error.rs:18:30
+  --> $DIR/struct-pattern-with-missing-fields-resolve-error.rs:22:30
    |
 LL |     if let Website { url, .. } = website {
    |            ------------------- this pattern doesn't include `title`, which is available in `Website`
 LL |         println!("[{}]({})", title, url);
    |                              ^^^^^ not found in this scope
 
-error: aborting due to 2 previous errors
+error[E0425]: cannot find value `a` in this scope
+  --> $DIR/struct-pattern-with-missing-fields-resolve-error.rs:28:20
+   |
+LL |     if let Foo::Bar { .. } = x {
+   |            --------------- this pattern doesn't include `a`, which is available in `Bar`
+LL |         println!("{a}");
+   |                    ^ help: a local variable with a similar name exists: `x`
+
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0425`.

